src/gtnh/gui.py

```diff
@@ -0,0 +1,390 @@
+
+    def __init__(self) -> None:
+        """
+        Constructor of the ArchivePopup class.
+
+        :return: None
+        """
+        tk.Toplevel.__init__(self)
+
+        # widgets on the window
+        self.progress_bar = Progressbar(self, orient="horizontal", mode="determinate", length=500)
+        self.progress_label = tk.Label(self, text="")
+        self.btn_start = tk.Button(self, text="start", command=self.start)
+
+        # grid manager
+        self.progress_bar.pack()
+        self.progress_label.pack()
+        self.btn_start.pack()
+
+    def start(self) -> None:
+        """
+        Method called when self.btn_start is pressed by the user. It starts the packaging process.
+
+        :return: None
+        """
+        github = Github(get_token())
+        organization = github.get_organization("GTNewHorizons")
+        gtnh_modpack = load_gtnh_manifest()
+        client_paths, server_paths = self.download_mods(gtnh_modpack, github, organization)
+
+        self.handle_pack_extra_files()
+        self.pack_client(client_paths)
+        self.pack_server(server_paths)
+        self.pack_technic()
+        self.make_deploader_json()
+        self.pack_curse()
+
+    def download_mods(self, gtnh_modpack: GTNHModpack, github: Github, organization: Organization) -> \
+            Tuple[List[AnyStr], List[AnyStr]]:
+        """
+        method to download all the mods required for the pack.
+
+        :param gtnh_modpack: GTNHModpack object. Represents the metadata of the modpack.
+        :param github: Github object.
+        :param organization: Organization object. Represent the GTNH organization.
+        :return: a list holding all the paths to the clientside mods and a list holding all the paths to the serverside
+                mod.
+        """
+        # computation of the progress per mod for the progressbar
+        delta_progress = 100 / len(gtnh_modpack.github_mods)
+
+        # lists holding the paths to the mods
+        client_paths = []
+        server_paths = []
+
+        # download of the mods
+        for mod in gtnh_modpack.github_mods:
+            # update progress bar
+            self.progress_bar["value"] += delta_progress
+            self.progress_bar["value"] = min(100.0, float(format(self.progress_bar["value"], ".2f")))
+            self.progress_label["text"] = f"downloading mods. Progress: {self.progress_bar['value']}%"
+            self.update()
+
+            # do the actual work
+            paths = download_mod(github, organization, mod)
+            if mod.side == "BOTH":
+                client_paths.extend(paths)
+                server_paths.extend(paths)
+            elif mod.side == "CLIENT":
+                client_paths.extend(paths)
+            elif mod.side == "SERVER":
+                server_paths.extend(paths)
+
+        # todo: make a similar thing for the curse mods
+
+        return client_paths, server_paths
+
+    def pack_client(self, client_paths: List[Path]) -> None:
```

Comment:
Not a requirement for this iteration, unless you want to, but we should split the actual building of the zip apart from the client usage of it.  If we want it to update the progress bar can add an optional callback to be utilized.
GUI.  As discussed in discord:
```python
def _pack_callback(self, label: str) -> None:

    # updating the progress bar
    self.progress_bar["value"] += self.delta_progress
    self.progress_bar["value"] = min(100.0, float(format(self.progress_bar["value"], ".2f")))
    self.progress_label["text"] = f"Packing client archive: {label}." \
                                  f"Progress: {self.progress_bar['value']}%"
    self.update()

def pack_client_gui(self, client_paths: List[Path]):
    # computation of the progress per mod for the progressbar
    self.delta_progress = 100 / len(client_paths)

    pack_client(client_paths, self._pack_callback)
```

Non GUI Package
```python
def pack_client(self, client_paths: List[Path], callback: Optional[Callable[...]] = None) -> None:
    """
    Method used to pack all the client files into a client archive.
    :param client_paths: a list containing all the Path objects refering to the files needed client side.
    :return: None
    """

    # remembering the cwd because it'll be changed during the zip operation
    cwd = os.getcwd()
    cache_dir = Path(ensure_cache_dir())
    os.chdir(cache_dir)

    # deleting any previous client archive
    if os.path.exists("client.zip"):
        os.remove("client.zip")
        print("previous client archive deleted")

    # zipping the files in the archive
    with ZipFile("client.zip", "w") as client_archive:
        for mod_path in client_paths:
            # writing the file in the zip
            client_archive.write(mod_path, mod_path.relative_to(cache_dir))
            if callback is not None:
                callback(mod_path.name)

    # restoring the cwd
    os.chdir(cwd)
    ```